### DISCUSSION TITLE
Why doesnt `useForm`  return the form values?

### DISCUSSION THREAD

**Discussion by crtl** [react-hook-form/react-hook-form#5596] [Q&A] [answered]
Currently the only ways to get a field value (as far as I know) are:
- using custom `onChange` handler
- using `watch("fieldName")`

Im not so experienced with React hooks and the underlying mechanics so im asking myself why
`useForm` doesnt return the form values in an object so we can easily access them like:

```typescript
type FormData {
  name: string;
}

const {register, formState, data} = useForm<FormData>();

console.log(data.name);
```

**Comment by bluebill1049** [comment]
then we will re-render your entire form/app on each `onChange`, it's one of the reasons why performance is important for any type of application. while we do offer the option to provide the form values on each on change, but not by default.

**Comment by rosemariesadler** [comment, marked as answer]
There are other ways to get a field value. `useForm` returns a function `getValues` that you can call to get the values of the form on demand. Additionally, if you do need to watch a single value in the form, due to performance issues with re-rendering the entire context on each change, it could be useful to create a new component that then uses `useWatch` for the single field that you need to watch. For example:

```js
export default function MyForm() {
  const { getValues, control, register, handleSubmit } = useForm();
  const onSubmit = (data) => console.log(data);
  const logValues = () => console.log(getValues());

  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      <input {...register("firstName")} placeholder="First Name" />
      <WatchFirstNameValue control={control} />
      <button onClick={logValues}>Log the Values</button>
      <input type="submit" />
    </form>
  );
}

const WatchFirstNameValue = () => {
  const firstNameValue = useWatch({ name: "firstName" });
  return `Hello ${firstNameValue}!`;
};

**Reply by bluebill1049** [reply]
thank you @rosemariesadler

**Comment by crtl** [comment]
@bluebill1049 @rosemariesadler Thank you both for your replies, this makes things clearer.
